### PR TITLE
First commit

### PR DESCRIPTION
**Pull request by pboothe**
Just a single function to find a list of files.  Its simplicity should help us get through a bunch of go readability issues really fast.

---
This change is [Reviewable](https://reviewable.io/reviews/m-lab/pusher/1)

### CODE REVIEW

File: findfiles_test.go [hunk 1]
```diff
@@ -0,0 +1,28 @@
+package pusher
```
Comment (gfr10598) [inline findfiles_test.go]: To encourage blackbox testing, I suggest putting tests in _test packages.  So try pusher_test.

Go tools recognize this and work nicely, but it restricts the tests to using the public interface.

File: findfiles.go [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+	"time"
+)
+
+// FindFiles recursively searches through a given directory to find all the files which are old enough to be eligible for upload.
+func FindFiles(directory string, minFileAge time.Duration) (error, []*LocalDataFile) {
+	eligibleFiles := make([]*LocalDataFile, 0)
+	eligibleTime := time.Now().Add(-minFileAge)
+	totalEligibleSize := int64(0)
+
+	err := filepath.Walk(directory, func(path string, info os.FileInfo, err error) error {
```
Comment (gfr10598) [inline findfiles.go]: Add a comment about the efficiency issue noted in the library documentation.  Would need to benchmark this, especially if we roll it out on old kernel.

Comment (pboothe) [reply findfiles.go]: They are worried about the overhead of `sort` on the filenames. I am not and I don't think we should be.

**Comment by gfr10598** [comment]
:lgtm:

---

Review status: 0 of 5 files reviewed at latest revision, 6 unresolved discussions.

---

*[findfiles.go, line 15 at r1](https://reviewable.io:443/reviews/m-lab/pusher/1#-KxhyJSlIBZxLfVrkwze:-KxhyJSlIBZxLfVrkwzf:bnkkj2c) ([raw file](https://github.com/m-lab/pusher/blob/7a66c20cca7ff5c7c97f768b0db3742db27ee179/findfiles.go#L15)):*
> ```Go
> )
> 
> // FindFiles recursively searches through a given directory to find all the files which are old enough to be eligible for upload.
> ```

Add to comment that result is sorted by mTime.

---

*[findfiles.go, line 17 at r1](https://reviewable.io:443/reviews/m-lab/pusher/1#-KxhvKvse1Qv6PKA7Z1g:-KxhvKvse1Qv6PKA7Z1h:b-n5422i) ([raw file](https://github.com/m-lab/pusher/blob/7a66c20cca7ff5c7c97f768b0db3742db27ee179/findfiles.go#L17)):*
> ```Go
> // FindFiles recursively searches through a given directory to find all the files which are old enough to be eligible for upload.
> func FindFiles(directory string, minFileAge time.Duration) (error, []*LocalDataFile) {
> 	eligibleFiles := make([]*LocalDataFile, 0)
> ```

FYI, []LocalDataFile is usually preferred, both here and in the return signature.  More often than not, pointers end up being counter productive, so usually, start with slice of objects, and profile if you think it might be important.
But I see below that you need to sort the array, so the pointers probably make sense.  Might be interesting to benchmark someday though.  8-)

---

*[findfiles.go, line 23 at r1](https://reviewable.io:443/reviews/m-lab/pusher/1#-KxhwJ9oTIi4Do_KZouR:-KxhwJ9oTIi4Do_KZouS:b-l4cvuw) ([raw file](https://github.com/m-lab/pusher/blob/7a66c20cca7ff5c7c97f768b0db3742db27ee179/findfiles.go#L23)):*
> ```Go
> 	err := filepath.Walk(directory, func(path string, info os.FileInfo, err error) error {
> 		if err != nil {
> 			return err
> ```

Comment what happens to returned errors?  This terminates the walk?

---

*[findfiles.go, line 32 at r1](https://reviewable.io:443/reviews/m-lab/pusher/1#-Kxhx-XUMUJ3htfY2AAm:-Kxhx-XUMUJ3htfY2AAn:b-6i1s20) ([raw file](https://github.com/m-lab/pusher/blob/7a66c20cca7ff5c7c97f768b0db3742db27ee179/findfiles.go#L32)):*
> ```Go
> 				fullRelativeName: path,
> 				info: info,
> 				cachedSize: info.Size(),
> ```

Nit: What is the point of cachedSize vs getting it from info?

---

*[findfiles.go, line 47 at r1](https://reviewable.io:443/reviews/m-lab/pusher/1#-KxhxEgKlrZRAEF59uWG:-KxhxEgLiaGmwaC7atK_:b-7mjqne) ([raw file](https://github.com/m-lab/pusher/blob/7a66c20cca7ff5c7c97f768b0db3742db27ee179/findfiles.go#L47)):*
> ```Go
> 	log.Printf("Total file count = %d", len(eligibleFiles))
> 	// Sort the files by mtime
> 	sort.Slice(eligibleFiles, func(i, j int) bool {
> ```

Ah - now I see why you want pointers.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/m-lab/pusher/1#-:-Kxhyb6XQbmBLG2LTQfE:bnfp4nl)*

**Comment by gfr10598** [comment]
Reviewed 5 of 5 files at r1.
Review status: all files reviewed at latest revision, 6 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/m-lab/pusher/1)*

File: findfiles_test.go (hunk 1 above)
Comment (pboothe) [reply findfiles_test.go]: Done

**Comment by pboothe** [comment]
Review status: all files reviewed at latest revision, 6 unresolved discussions.

---

*[findfiles.go, line 15 at r1](https://reviewable.io:443/reviews/m-lab/pusher/1#-KxhyJSlIBZxLfVrkwze:-Kxi5KCyhDzXOS9cNTLd:b-896fix) ([raw file](https://github.com/m-lab/pusher/blob/7a66c20cca7ff5c7c97f768b0db3742db27ee179/findfiles.go#L15)):*
<details><summary><i>Previously, gfr10598 (Gregory Russell) wrote…</i></summary><blockquote>

Add to comment that result is sorted by mTime.
</blockquote></details>

Done.

---

*[findfiles.go, line 17 at r1](https://reviewable.io:443/reviews/m-lab/pusher/1#-KxhvKvse1Qv6PKA7Z1g:-Kxi5LIfwL80I0PBw5Dz:bb6sh2i) ([raw file](https://github.com/m-lab/pusher/blob/7a66c20cca7ff5c7c97f768b0db3742db27ee179/findfiles.go#L17)):*
<details><summary><i>Previously, gfr10598 (Gregory Russell) wrote…</i></summary><blockquote>

FYI, []LocalDataFile is usually preferred, both here and in the return signature.  More often than not, pointers end up being counter productive, so usually, start with slice of objects, and profile if you think it might be important.
But I see below that you need to sort the array, so the pointers probably make sense.  Might be interesting to benchmark someday though.  8-)

</blockquote></details>

Ack.

---

*[findfiles.go, line 23 at r1](https://reviewable.io:443/reviews/m-lab/pusher/1#-KxhwJ9oTIi4Do_KZouR:-Kxi5Mf6aekbZ--iPodt:b-896fix) ([raw file](https://github.com/m-lab/pusher/blob/7a66c20cca7ff5c7c97f768b0db3742db27ee179/findfiles.go#L23)):*
<details><summary><i>Previously, gfr10598 (Gregory Russell) wrote…</i></summary><blockquote>

Comment what happens to returned errors?  This terminates the walk?
</blockquote></details>

Done.

---

*[findfiles.go, line 32 at r1](https://reviewable.io:443/reviews/m-lab/pusher/1#-Kxhx-XUMUJ3htfY2AAm:-Kxi5NKUbMyRX2O6l0q8:bbnq2sa) ([raw file](https://github.com/m-lab/pusher/blob/7a66c20cca7ff5c7c97f768b0db3742db27ee179/findfiles.go#L32)):*
<details><summary><i>Previously, gfr10598 (Gregory Russell) wrote…</i></summary><blockquote>

Nit: What is the point of cachedSize vs getting it from info?
</blockquote></details>

Added a comment to the struct:
```
        // Cache the size so sums of filesize are always using the same
        // integer value, in case the file grows in size between walk and
        // upload.```
```


---

*[findfiles.go, line 47 at r1](https://reviewable.io:443/reviews/m-lab/pusher/1#-KxhxEgKlrZRAEF59uWG:-Kxi5WUAFbLuAfjYH2mI:bb6sh2i) ([raw file](https://github.com/m-lab/pusher/blob/7a66c20cca7ff5c7c97f768b0db3742db27ee179/findfiles.go#L47)):*
<details><summary><i>Previously, gfr10598 (Gregory Russell) wrote…</i></summary><blockquote>

Ah - now I see why you want pointers.
</blockquote></details>

Ack.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/m-lab/pusher/1)*

**Pull request merged**